File: curzon_numbers/curzon_numbers.cpp
#include <iomanip>
#include <iostream>
#include <vector>

#include <gmpxx.h>

bool is_curzon(int n, int k) {
    mpz_class p;
    mpz_ui_pow_ui(p.get_mpz_t(), k, n);
    return (p + 1) % (k * n + 1) == 0;
}

int main() {
    for (int k = 2; k <= 10; k += 2) {
        std::cout << "Curzon numbers with base " << k << ":\n";
        int count = 0, n = 1;
        for (; count < 50; ++n) {
            if (is_curzon(n, k)) {
                std::cout << std::setw(4) << n
                          << (++count % 10 == 0 ? '\n' : ' ');
            }
        }
        for (;;) {
            if (is_curzon(n, k))
                ++count;
            if (count == 1000)
                break;
            ++n;
        }
        std::cout << "1000th Curzon number with base " << k << ": " << n
                  << "\n\n";
    }
    return 0;
}

File: curzon_numbers/curzon_numbers.rs
// [dependencies]
// rug = "1.15.0"

fn is_curzon(n: u32, k: u32) -> bool {
    use rug::{Complete, Integer};
    (Integer::u_pow_u(k, n).complete() + 1) % (k * n + 1) == 0
}

fn main() {
    for k in (2..=10).step_by(2) {
        println!("Curzon numbers with base {k}:");
        let mut count = 0;
        let mut n = 1;
        while count < 50 {
            if is_curzon(n, k) {
                count += 1;
                print!("{:4}{}", n, if count % 10 == 0 { "\n" } else { " " });
            }
            n += 1;
        }
        loop {
            if is_curzon(n, k) {
                count += 1;
                if count == 1000 {
                    break;
                }
            }
            n += 1;
        }
        println!("1000th Curzon number with base {k}: {n}\n");
    }
}
